packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/FileSelectorDelegate.java

```diff
@@ -0,0 +1,217 @@
+ */
+public class FileSelectorDelegate
+    implements PluginRegistry.ActivityResultListener,
+        PluginRegistry.RequestPermissionsResultListener {
+  static final int REQUEST_CODE_GET_DIRECTORY_PATH = 2342;
+  static final int REQUEST_CODE_OPEN_FILE = 2343;
+
+  static String cacheFolder = "file_selector";
+
+  Intent openFileIntent = new Intent();
+
+  private Messages.Result pendingResult;
+  private final Activity activity;
+
+  @Override
+  public boolean onRequestPermissionsResult(
+      int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
+    return true;
+  }
+
+  public FileSelectorDelegate(final Activity activity) {
+    this(activity, null);
+  }
+
+  /**
+   * These constructors are used exclusively for testing; they can be used to provide mocks to final
+   * fields of this class. Otherwise, those fields would have to be mutable and visible.
+   */
+  @VisibleForTesting
+  FileSelectorDelegate() {
+    this(null, null);
+  }
+
+  @VisibleForTesting
+  FileSelectorDelegate(final Activity activity, final Messages.Result result) {
+    this.activity = activity;
+    this.pendingResult = result;
+  }
+
+  public void clearCache() {
+    PathUtils.clearCache(this.activity, cacheFolder);
+  }
+
+  public void getDirectoryPath(@Nullable String initialDirectory, Messages.Result<String> result) {
+    if (setPendingResult(result)) {
+      finishWithAlreadyActiveError(result);
+      return;
+    }
+
+    launchGetDirectoryPath(initialDirectory);
+  }
+
+  public void openFile(
+      @NonNull Messages.SelectionOptions options, Messages.Result<List<String>> result) {
+    if (setPendingResult(result)) {
+      finishWithAlreadyActiveError(result);
+      return;
+    }
+
+    Boolean multipleFiles = options.getAllowMultiple();
+    List<String> acceptedTypeGroups = options.getAllowedTypes();
+
+    launchOpenFile(multipleFiles, acceptedTypeGroups);
+  }
+
+  void launchGetDirectoryPath(@Nullable String initialDirectory) {
+    Intent getDirectoryPathIntent = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
+
+    if (initialDirectory != null && !initialDirectory.isEmpty()) {
+      Uri uri = getDirectoryPathIntent.getParcelableExtra("android.provider.extra.INITIAL_URI");
+      String scheme = uri.toString();
+      scheme = scheme.replace("/root/", initialDirectory);
+      uri = Uri.parse(scheme);
+      getDirectoryPathIntent.putExtra("android.provider.extra.INITIAL_URI", uri);
+    }
+    activity.startActivityForResult(getDirectoryPathIntent, REQUEST_CODE_GET_DIRECTORY_PATH);
+  }
+
+  @Override
+  public boolean onActivityResult(int requestCode, int resultCode, Intent data) {
+    switch (requestCode) {
+      case REQUEST_CODE_GET_DIRECTORY_PATH:
+        handleGetDirectoryPathResult(resultCode, data);
+        break;
+      case REQUEST_CODE_OPEN_FILE:
+        handleOpenFileResult(resultCode, data);
+        break;
+      default:
+        return false;
```

Comment:
Well It is expected to this code won't be reached because there are only two `startActivityForResult` calls, one for each pre-defined request code: `REQUEST_CODE_GET_DIRECTORY_PATH` and `REQUEST_CODE_OPEN_FILE`. Thoughts?

Reply:
If it's supposed to be unreachable, then it should probably throw instead of silently failing.